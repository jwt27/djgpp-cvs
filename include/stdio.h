/* Copyright (C) 2014 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2013 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2001 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2000 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1998 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1995 DJ Delorie, see COPYING.DJ for details */

#ifndef __dj_include_stdio_h_
#define __dj_include_stdio_h_

#ifdef __cplusplus
extern "C" {
#endif

#ifndef __dj_ENFORCE_ANSI_FREESTANDING

#include <sys/version.h>
#include <sys/djtypes.h>
  
#define _IOFBF    	00001  /*  File is full buffered.  */
#define _IONBF    	00002  /*  File is not buffered.  */
#define _IOLBF    	00004  /*  File is line buffered.  */

/* Some programs think they know better... */
#undef NULL
#if (__GNUC__ >= 4) && defined(__cplusplus)
#  define NULL          __null
#elif defined(__cplusplus)
#  define NULL          0
#else
#  define NULL          ((void*)0)
#endif

#define BUFSIZ		16384
#define EOF		(-1)
#define FILENAME_MAX	260
#define FOPEN_MAX	20
#define L_tmpnam	260
#define TMP_MAX		999999

#define SEEK_SET	0
#define SEEK_CUR	1
#define SEEK_END	2

#ifndef _VA_LIST
__DJ_va_list
#define _VA_LIST
#endif

#ifndef _SIZE_T
__DJ_size_t
#define _SIZE_T
#endif

#ifndef _SSIZE_T
__DJ_ssize_t
#define _SSIZE_T
#endif

/* Note that the definitions of these fields are NOT guaranteed!  They
   may change with any release without notice!  The fact that they
   are here at all is to comply with ANSI specifictions. */
   
typedef struct {
  ssize_t  _cnt;
  char    *_ptr;
  char    *_base;
  size_t   _bufsiz;
  int      _flag;
  int      _file;
  char    *_name_to_remove;
  size_t   _fillsize;
} FILE;

typedef unsigned long		fpos_t;

extern FILE __dj_stdin, __dj_stdout, __dj_stderr;
#define stdin	(&__dj_stdin)
#define stdout	(&__dj_stdout)
#define stderr	(&__dj_stderr)

void	clearerr(FILE *_stream);
int	fclose(FILE *_stream);
int	feof(FILE *_stream);
int	ferror(FILE *_stream);
int	fflush(FILE *_stream);
int	fgetc(FILE *_stream);
int	fgetpos(FILE *_stream, fpos_t *_pos);
char *	fgets(char *_s, int _n, FILE *_stream);
FILE *	fopen(const char *_filename, const char *_mode);
int	fprintf(FILE *_stream, const char *_format, ...);
int	fputc(int _c, FILE *_stream);
int	fputs(const char *_s, FILE *_stream);
size_t	fread(void *_ptr, size_t _size, size_t _nelem, FILE *_stream);
FILE *	freopen(const char *_filename, const char *_mode, FILE *_stream);
int	fscanf(FILE *_stream, const char *_format, ...);
int	fseek(FILE *_stream, long _offset, int _mode);
int	fsetpos(FILE *_stream, const fpos_t *_pos);
long	ftell(FILE *_stream);
size_t	fwrite(const void *_ptr, size_t _size, size_t _nelem, FILE *_stream);
int	getc(FILE *_stream);
int	getchar(void);
char *	gets(char *_s);
void	perror(const char *_s);
int	printf(const char *_format, ...);
int	putc(int _c, FILE *_stream);
int	putchar(int _c);
int	puts(const char *_s);
int	remove(const char *_filename);
int	rename(const char *_old, const char *_new);
void	rewind(FILE *_stream);
int	scanf(const char *_format, ...);
void	setbuf(FILE *_stream, char *_buf);
int	setvbuf(FILE *_stream, char *_buf, int _mode, size_t _size);
int	sprintf(char *_s, const char *_format, ...);
int	sscanf(const char *_s, const char *_format, ...);
FILE *	tmpfile(void);
char *	tmpnam(char *_s);
int	ungetc(int _c, FILE *_stream);
int	vfprintf(FILE *_stream, const char *_format, va_list _ap);
int	vprintf(const char *_format, va_list _ap);
int	vsprintf(char *_s, const char *_format, va_list _ap);

#if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L) \
  || !defined(__STRICT_ANSI__) || defined(__cplusplus)

int	snprintf(char *str, size_t n, const char *fmt, ...);
int	vfscanf(FILE *_stream, const char *_format, va_list _ap);
int	vscanf(const char *_format, va_list _ap);
int	vsnprintf(char *str, size_t n, const char *fmt, va_list ap);
int	vsscanf(const char *_s, const char *_format, va_list _ap);

#endif /* (__STDC_VERSION__ >= 199901L) || !__STRICT_ANSI__ */

#ifndef __STRICT_ANSI__

#define L_ctermid 20
#define L_cusrid
/* #define STREAM_MAX	20 - DOS can change this */

int	dprintf(int _fd, const char *_format, ...) __attribute__ ((__format__ (__printf__, 2, 3)));
int	fileno(FILE *_stream);
FILE *	fdopen(int _fildes, const char *_type);
int	mkstemp(char *_template);
int	pclose(FILE *_pf);
FILE *	popen(const char *_command, const char *_mode);
char *	tempnam(const char *_dir, const char *_prefix);
int	vdprintf(int _fd, const char *_format, va_list _ap) __attribute__ ((__format__ (__printf__, 2, 0)));

#ifndef _POSIX_SOURCE

extern FILE __dj_stdprn, __dj_stdaux;
#define stdprn	(&__dj_stdprn)
#define stdaux	(&__dj_stdaux)

#define P_tmpdir "c:/"

void	_djstat_describe_lossage(FILE *_to_where);
int	_doprnt(const char *_fmt, va_list _args, FILE *_f);
int	_doscan(FILE *_f, const char *_fmt, va_list _args);
int	_doscan_low(FILE *, int (*)(FILE *_get), int (*_unget)(int, FILE *), const char *_fmt, va_list _args);
int	fpurge(FILE *_f);
int	getw(FILE *_f);
char *	mktemp(char *_template);
int	putw(int _v, FILE *_f);
void	setbuffer(FILE *_f, void *_buf, int _size);
void	setlinebuf(FILE *_f);
int	_rename(const char *_old, const char *_new);	/* Simple (no directory) */
int	asprintf(char **_sp, const char *_format, ...) __attribute__((format (__printf__, 2, 3)));
char *	asnprintf(char *_s, size_t *_np, const char *_format, ...) __attribute__((format (__printf__, 3, 4)));
int	vasprintf(char **_sp, const char *_format, va_list _ap) __attribute__((format (__printf__, 2, 0)));
char *	vasnprintf(char *_s, size_t *_np, const char *_format, va_list _ap) __attribute__((format (__printf__, 3, 0)));

#ifndef _OFF_T
__DJ_off_t
#define _OFF_T
#endif
#ifndef _OFF64_T
__DJ_off64_t
#define _OFF64_T
#endif
int	fseeko(FILE *_stream, off_t _offset, int _mode);
off_t	ftello(FILE *_stream);
int	fseeko64(FILE *_stream, off64_t _offset, int _mode);
off64_t	ftello64(FILE *_stream);

#endif /* !_POSIX_SOURCE */
#endif /* !__STRICT_ANSI__ */
#endif /* !__dj_ENFORCE_ANSI_FREESTANDING */

#ifndef __dj_ENFORCE_FUNCTION_CALLS
#endif /* !__dj_ENFORCE_FUNCTION_CALLS */

#ifdef __cplusplus
}
#endif

#endif /* !__dj_include_stdio_h_ */
