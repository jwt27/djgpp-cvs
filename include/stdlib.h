/* Copyright (C) 2015 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2014 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2013 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2001 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2000 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1999 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1998 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1995 DJ Delorie, see COPYING.DJ for details */
#ifndef __dj_include_stdlib_h_
#define __dj_include_stdlib_h_

#ifdef __cplusplus
extern "C" {
#endif

#ifndef __dj_ENFORCE_ANSI_FREESTANDING

#include <sys/djtypes.h>
  
/* Some programs think they know better... */
#undef NULL
#if (__GNUC__ >= 4) && defined(__cplusplus)
#  define NULL          __null
#elif defined(__cplusplus)
#  define NULL          0
#else
#  define NULL          ((void*)0)
#endif

#define EXIT_FAILURE	1
#define EXIT_SUCCESS	0
#define MB_CUR_MAX	__dj_mb_cur_max
#define RAND_MAX	2147483647

extern int __dj_mb_cur_max;

typedef struct {
  int quot;
  int rem;
} div_t;

typedef struct {
  long quot;
  long rem;
} ldiv_t;

#ifndef _SIZE_T
__DJ_size_t
#define _SIZE_T
#endif

#ifndef _WCHAR_T
__DJ_wchar_t
#define _WCHAR_T
#endif

void	abort(void) __attribute__((noreturn));
int	abs(int _i);
int	atexit(void (*_func)(void));
double	atof(const char *_s);
int	atoi(const char *_s);
long	atol(const char *_s);
void *	bsearch(const void *_key, const void *_base, size_t _nelem,
		size_t _size, int (*_cmp)(const void *_ck, const void *_ce));
void *	calloc(size_t _nelem, size_t _size);
div_t	div(int _numer, int _denom);
void	exit(int _status) __attribute__((noreturn));
void	free(void *_ptr);
char *	getenv(const char *_name);
long	labs(long _i);
ldiv_t	ldiv(long _numer, long _denom);
void *	malloc(size_t _size);
int	mblen(const char *_s, size_t _n);
size_t	mbstowcs(wchar_t *_wcs, const char *_s, size_t _n);
int	mbtowc(wchar_t *_pwc, const char *_s, size_t _n);
void	qsort(void *_base, size_t _nelem, size_t _size,
	      int (*_cmp)(const void *_e1, const void *_e2));
int	rand(void);
void *	realloc(void *_ptr, size_t _size);
void	srand(unsigned _seed);
double	strtod(const char *_s, char **_endptr);
long	strtol(const char *_s, char **_endptr, int _base);
unsigned long	strtoul(const char *_s, char **_endptr, int _base);
int	system(const char *_s);
size_t	wcstombs(char *_s, const wchar_t *_wcs, size_t _n);
int	wctomb(char *_s, wchar_t _wchar);

#if (defined(__STDC_VERSION__) && __STDC_VERSION__ >= 199901L) \
  || !defined(__STRICT_ANSI__) || defined(__cplusplus)

typedef struct {
  long long int quot;
  long long int rem;
} lldiv_t;

void		_Exit(int _status) __attribute__((noreturn));
long long int	atoll(const char *_s);
long long int	llabs(long long int _i);
lldiv_t		lldiv(long long int _numer, long long int _denom);
float		strtof(const char *_s, char **_endptr);
long double	strtold(const char *_s, char **_endptr);
long long int	strtoll(const char *_s, char **_endptr, int _base);
unsigned long long int	strtoull(const char *_s, char **_endptr, int _base);

#endif /* (__STDC_VERSION__ >= 199901L) || !__STRICT_ANSI__ */

#ifndef __STRICT_ANSI__

long	a64l(const char *_string);
char *	l64a(long _value);
char *	mktemp(char *_template);
char *	mkdtemp(char *_template);
int	mkstemp(char *_template);
int	putenv(char *_val);
char *	realpath(const char *_path, char *_resolved);
int	setenv(const char *_var, const char *_val, int _overwrite);
int	unsetenv(const char *_var);

#ifndef _POSIX_SOURCE

void *		alloca(size_t _size);
long double	_atold(const char *_s);
void		cfree(void *_ptr);
double          drand48(void);
char *		ecvtbuf(double _val, int _nd, int *_dp, int *_sn, char *_bf);
char *		ecvt(double _val, int _nd, int *_dp, int *_sn);
double          erand48(unsigned short _state[3]);
char *		fcvtbuf(double _val, int _nd, int *_dp, int *_sn, char *_bf);
char *		fcvt(double _val, int _nd, int *_dp, int *_sn);
char *		gcvt(double _val, int _nd, char *_buf);
char *		getpass(const char *_prompt);
int		getlongpass(const char *_prompt, char *_buffer, int _max_len);
char *		itoa(int _value, char *_buffer, int _radix);
long            jrand48(unsigned short _state[3]);
void            lcong48(unsigned short _param[7]);
unsigned long   lrand48(void);
long            mrand48(void);
unsigned long   nrand48(unsigned short _state[3]);
unsigned short *seed48(unsigned short _state_seed[3]);
void            srand48(long _seedval);
int		stackavail(void);
long double	_strtold(const char *_s, char **_endptr);
void		swab(const void *_from, void *_to, int _nbytes);
void *		valloc (size_t _amt);

#ifndef alloca
#define alloca __builtin_alloca
#endif

/* BSD Random Number Generator */
char  *	initstate (unsigned _seed, char *_arg_state, int _n);
char  *	setstate(char *_arg_state);
long	random(void);
int	srandom(int _seed);

/* These vary in expected prototype, so we just don't prototype them.
void		xfree(void *_ptr);
void *		xmalloc(size_t _size);
void *		xrealloc(void *ptr, size_t _size);
*/

#define __system_redirect	      0x0001 /* redirect internally */
#define __system_call_cmdproc	      0x0002 /* always call COMMAND/$SHELL */
#define __system_use_shell	      0x0004 /* use $SHELL if set */
#define __system_allow_multiple_cmds  0x0008 /* allow `cmd1; cmd2; ...' */
#define __system_allow_long_cmds      0x0010 /* handle commands > 126 chars  */
#define __system_emulate_command      0x0020 /* try to emulate the shell */
#define __system_handle_null_commands 0x1000 /* ignore cmds with no effect */
#define __system_ignore_chdir	      0x2000 /* make `cd' be a null command */
#define __system_emulate_chdir	      0x4000 /* handle `cd' internally */

extern int __system_flags;

#endif /* !_POSIX_SOURCE */
#endif /* !__STRICT_ANSI__ */
#endif /* !__dj_ENFORCE_ANSI_FREESTANDING */

#ifndef __dj_ENFORCE_FUNCTION_CALLS
#endif /* !__dj_ENFORCE_FUNCTION_CALLS */

#ifdef __cplusplus
}
#endif

#endif /* !__dj_include_stdlib_h_ */
