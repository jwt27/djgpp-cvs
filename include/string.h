/* Copyright (C) 2017 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2013 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2003 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2002 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2000 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1998 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1995 DJ Delorie, see COPYING.DJ for details */
#ifndef __dj_include_string_h_
#define __dj_include_string_h_

#include <stddef.h>

int	strcoll(const char *_s1, const char *_s2);
char *	strerror(int _errcode);
size_t	strxfrm(char *  _s1, const char *  _s2, size_t _n);

//int	strerror_r(int _errnum, char *_strerrbuf, size_t _buflen);

int	memicmp(const void *_s1, const void *_s2, size_t _n);
char *	strdup(const char *_s);
char *	strndup(const char *_s, size_t _n);
size_t	strlcat(char *_dest, const char *_src, size_t _size);
size_t	strlcpy(char *_dest, const char *_src, size_t _size);
char *	strlwr(char *_s);
int	strcasecmp(const char *_s1, const char *_s2);
int	stricmp(const char *_s1, const char *_s2);
int	strncasecmp(const char *_s1, const char *_s2, size_t _n);
int	strnicmp(const char *_s1, const char *_s2, size_t _n);
char *	strupr(char *_s);

#define __USE_GNU
#include_next <string.h>
#undef __USE_GNU
#ifdef __USE_POSIX_IMPLICITLY
#undef __USE_POSIX_IMPLICITLY
#undef _DEFAULT_SOURCE
#undef _POSIX_SOURCE
#endif

#endif /* !__dj_include_string_h_ */
