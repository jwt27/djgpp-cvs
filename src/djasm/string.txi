@c Intel mnemonics
@opindex MOVSB
@opindex MOVSW
@opindex MOVSD
@opindex CMPSB
@opindex CMPSW
@opindex CMPSD
@opindex SCASB
@opindex SCASW
@opindex SCASD
@opindex LODSB
@opindex LODSW
@opindex LODSD
@opindex STOSB
@opindex STOSW
@opindex STOSD
@opindex REP
@opindex REPE
@opindex REPZ
@opindex REPNE
@opindex REPNZ
@opindex CLD
@opindex CLD

@c opcode description
@opindex Move string (byte)
@opindex Move string (word)
@opindex Move string (double word)
@opindex Compare strings (byte)
@opindex Compare strings (word)
@opindex Compare strings (double word)
@opindex Scan string (byte)
@opindex Scan string (word)
@opindex Scan string (double word)
@opindex Load string (byte)
@opindex Load string (word)
@opindex Load string (double word)
@opindex Store string (byte)
@opindex Store string (word)
@opindex Store string (double word)
@opindex Repeat while not end of string
@opindex Repeat while not end of string and strings are equal
@opindex Repeat while not end of string and strings are equal
@opindex Repeat while not end of string and strings are not equal
@opindex Repeat while not end of string and strings are not equal
@opindex Clear DF
@opindex Set DF

@c djasm mnemonics
@opindex movsb
@opindex movsw
@opindex movsd
@opindex cmpsb
@opindex cmpsw
@opindex cmpsd
@opindex scasb
@opindex scasw
@opindex scasd
@opindex lodsb
@opindex lodsw
@opindex lodsd
@opindex stosb
@opindex stosw
@opindex stosd
@opindex rep
@opindex repe
@opindex repz
@opindex repne
@opindex repnz
@opindex cld
@opindex std

@c The byte, word, and double word versions of some string mnemonics
@c distinguished in my Intel reference, so I left the djasm mnemonics
@c distinguished as well.

@multitable {--------} {----------} {--------------------------------------------------}
@item Intel     @tab DJASM      @tab Meaning
@item ~~~~~~~~  @tab ~~~~~~~~~~ @tab ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
@item MOVSB     @tab movsb      @tab Move string (byte)
@item MOVSW     @tab movsw      @tab Move string (word)
@item MOVSD     @tab movsd      @tab Move string (double word)
@item
@item CMPSB     @tab cmpsb      @tab Compare strings (byte)
@item CMPSW     @tab cmpsw      @tab Compare strings (word)
@item CMPSD     @tab cmpsd      @tab Compare strings (double word)
@item
@item SCASB     @tab scasb      @tab Scan string (byte)
@item SCASW     @tab scasw      @tab Scan string (word)
@item SCASD     @tab scasd      @tab Scan string (double word)
@item
@item LODSB     @tab lodsb      @tab Load string (byte)
@item LODSW     @tab lodsw      @tab Load string (word)
@item LODSD     @tab lodsd      @tab Load string (double word)
@item
@item STOSB     @tab stosb      @tab Store string (byte)
@item STOSW     @tab stosw      @tab Store string (word)
@item STOSD     @tab stosd      @tab Store string (double word)
@item
@item REP       @tab rep        @tab MOVS/STOS: Repeat while not end of string
@item REPE      @tab repe       @tab CMPS/SCAS: Repeat while not end of string and strings are equal
@item REPZ      @tab repz       @tab CMPS/SCAS: Repeat while not end of string and strings are equal
@item REPNE     @tab repne      @tab CMPS/SCAS: Repeat while not end of string and strings are not equal
@item REPNZ     @tab repnz      @tab CMPS/SCAS: Repeat while not end of string and strings are not equal
@item
@item CLD       @tab cld        @tab Clear DF
@item CLD       @tab std        @tab Set DF
@end multitable
