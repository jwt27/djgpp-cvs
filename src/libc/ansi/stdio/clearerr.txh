@node clearerr, stdio
@findex clearerr
@subheading Syntax

@example
#include <stdio.h>

void clearerr(FILE *stream);
@end example

@subheading Description

This function clears the EOF and error indicators for the file
@var{stream}. 

@subheading Return Value

None.

@subheading Portability

@portability ansi, posix

@subheading Example

@example
clearerr(stdout);
@end example

