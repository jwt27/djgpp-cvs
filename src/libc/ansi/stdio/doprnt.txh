@node _doprnt, stdio
@findex _doprnt
@subheading Syntax

@example
#include <stdio.h>

int _doprnt(const char *format, void *params, FILE *file);
@end example

@subheading Description

This is an internal function that is used by all the @code{printf} style
functions, which simply pass their format, arguments, and stream to this
function. 

@xref{printf}, for a discussion of the allowed formats and arguments. 

@subheading Return Value

The number of characters generated is returned.

@subheading Portability

@portability !ansi, !posix

@subheading Example

@example
int args[] = @{ 1, 2, 3, 66 @};
_doprnt("%d %d %d %c\n", args, stdout);
@end example

