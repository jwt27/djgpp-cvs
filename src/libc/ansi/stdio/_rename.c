/* Copyright (C) 2011 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2001 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 2000 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1999 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1998 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1997 DJ Delorie, see COPYING.DJ for details */
/* Copyright (C) 1995 DJ Delorie, see COPYING.DJ for details */
#include <libc/stubs.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include <unistd.h>
#include <fcntl.h>
#include <sys/stat.h>
#include <io.h>
#include <dos.h>
#include <go32.h>
#include <dpmi.h>
#include <libc/dosio.h>
 
int _rename(const char *old, const char *new)
{
  __dpmi_regs r;
  int i;
  int olen    = strlen(old) + 1;
  int use_lfn = _USE_LFN;
  char tempfile[FILENAME_MAX], tempfile1[FILENAME_MAX];
  const char *orig = old;
  int lfn_fd = -1;
  int identical_but_for_case = 0;

  if (__file_exists(new)
      && strcmp(_truename(old, tempfile), _truename(new, tempfile1)) == 0)
    {
      /* NEW might seem to exist because the filesystem is case-
	 insensitive, if OLD and NEW are identical but for the
	 letter case.  Allow renaming `foo' into `Foo' under LFN.  */
      if (use_lfn)
	{
	  const char *s1 = old + strlen(old), *s2 = new + strlen(new);

	  /* Find the basename of both OLD and NEW.  */
	  while (s1 > old && s2 > new
		 && strchr(":/\\", s1[-1]) == NULL
		 && strchr(":/\\", s2[-1]) == NULL)
	  {
	    s1--;
	    s2--;
	  }
	  /* If OLD and NEW are really the same file, do nothing.  */
	  if (strcmp(s1, s2) == 0)
	    return 0;
	  identical_but_for_case = 1;
	}
      else
	return 0;  /* no LFN; foo and Foo are *always* the same file */
    }

  r.x.dx = __tb_offset;
  r.x.di = __tb_offset + olen;
  r.x.ds = r.x.es = __tb_segment;

  if (use_lfn && !identical_but_for_case)
  {
    /* Windows 95 bug: for some filenames, when you rename
       file -> file~ (as in Emacs, to leave a backup), the
       short 8+3 alias doesn't change, which effectively
       makes OLD and NEW the same file.  We must rename
       through a temporary file to work around this.  */

    char *pbase = 0, *p;
    static char try_char[] = "abcdefghijklmnopqrstuvwxyz012345789";
    int idx = sizeof(try_char) - 1;

    /* Generate a temporary name.  Can't use `tmpnam', since $TMPDIR
       might point to another drive, which will fail the DOS call.  */
    strcpy(tempfile, old);
    for (p = tempfile; *p; p++)	/* ensure temporary is on the same drive */
      if (*p == '/' || *p == '\\' || *p == ':')
	pbase = p;
    if (pbase)
      pbase++;
    else
      pbase = tempfile;
    strcpy(pbase, "X$$djren$$.$$temp$$");

    do
    {
      if (idx <= 0)
	return -1;
      *pbase = try_char[--idx];
    } while (_chmod(tempfile, 0) != -1);

    r.x.flags = 1;  /* Always set CF before calling a 0x71NN function. */
    r.x.ax = 0x7156;
    _put_path2(tempfile, olen);
    _put_path(old);
    __dpmi_int(0x21, &r);
    if ((r.x.flags & 1) || (r.x.ax == 0x7100))
    {
      /*  Never assume that the complete LFN API is implemented,
          so check that AX != 0x7100.  E.G.: MSDOS 6.22 and DOSLFN 0.40.
          If not supported fail.  */
      errno = __doserr_to_errno(r.x.ax);
      return -1;
    }

    /* Now create a file with the original name.  This will
       ensure that NEW will always have a 8+3 alias
       different from that of OLD.  (Seems to be required
       when NameNumericTail in the Registry is set to 0.)  */
    lfn_fd = _creat(old, 0);

    olen = strlen(tempfile) + 1;
    old  = tempfile;
    r.x.di = __tb_offset + olen;
  }

  for (i = 0; i < 2; i++)
  {
    if (use_lfn)
    {
      r.x.flags |= 1;  /* Always set CF before calling a 0x71NN function. */
      r.x.ax = 0x7156;
    }
#if 0
    /* It seems that no version of DOS, including DOS 8, which is part
       of Windows/ME, implements this function.  Without LFN, this fails
       _rename on Windows/ME.  Disabled.  */
    else if ((_osmajor > 7 && _osmajor < 10) /* OS/2 returns v10 and above */
	     || (_osmajor == 7 && _osminor >= 20))
    {
      /* DOS 7.20 (Windows 98) and later supports a new function with
	 a maximum path length of 128 characters instead of 67.  This
	 is important for deeply-nested directories.  */
      r.x.ax = 0x43ff;
      r.x.bp = 0x5053;
      r.h.cl = 0x56;
    }
#endif
    else
      r.h.ah = 0x56;
    _put_path2(new, olen);
    _put_path(old);
    __dpmi_int(0x21, &r);
    if ((r.x.flags & 1) || (r.x.ax == 0x7100))
    {
      /*  Never assume that the complete LFN API is implemented,
          so check that AX != 0x7100.  E.G.: MSDOS 6.22 and DOSLFN 0.40.
          If not supported fail.  */

      if (i == 0
	  && !identical_but_for_case /* don't nuke OLD! */
	  && (r.x.ax == 5 || (r.x.ax == 2 && __file_exists(old))
	      /* Windows 2000 returns B7h when the target file exists.  */
	      || r.x.ax == 0xb7))
	remove(new);		 /* and try again */
      else
      {
	errno = __doserr_to_errno(r.x.ax);

	/* Restore to original name if we renamed it to temporary.  */
	if (use_lfn && !identical_but_for_case)
	{
	  if (lfn_fd != -1)
	  {
	    _close (lfn_fd);
	    remove (orig);
	  }
	  _put_path2(orig, olen);
	  _put_path(tempfile);
	  r.x.ax = 0x7156;
	  __dpmi_int(0x21, &r);
	}
	return -1;
      }
    }
    else
      break;
  }

  /* Success.  Delete the file possibly created to work
     around the Windows 95 bug.  */
  if (lfn_fd != -1)
    return (_close (lfn_fd) == 0) ? remove (orig) : -1;
  return 0;
}

#ifdef TEST

#include <string.h>

int main(int argc, char *argv[])
{
  if (argc > 2)
    {
      printf ("%s -> %s: ", argv[1], argv[2]);
      if (_rename (argv[1], argv[2]))
	printf ("%s\n", strerror (errno));
      else
	printf ("Done\n");
    }
  return 0;
}

#endif
