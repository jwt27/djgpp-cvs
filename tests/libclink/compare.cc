/*
**  Usage: compare oldlib.armap newlib.armap
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "slist.h"

void
loadsyms(char *fn, StringList &sl)
{
  FILE *f;
  char line[1000];
  char sym[1000];
  f = fopen(fn, "r");
  if (f == 0)
  {
    perror(fn);
    exit(1);
  }
  fgets(line, 1000, f);
  fgets(line, 1000, f);
  while (fgets(line, 1000, f))
  {
    if (line[0] <= ' ')
      break;
    sscanf(line, "%s", sym);
    sl.add(sym);
  }
  fclose(f);
  sl.sort();
}

int
main(int argc, char **argv)
{
  StringList old_objs;
  StringList new_objs;
  int i, col=80, need_banner=1;

  loadsyms(argv[1], old_objs);
  loadsyms(argv[2], new_objs);

  for (i=0; i<old_objs.count; i++)
  {
    char *s = old_objs[i];
    if (s[0] != '_')
      continue;
    if (!new_objs.has(s))
    {
      if (need_banner)
      {
	need_banner = 0;
	printf("Symbols in old library, but not in new one:");
      }
      col += strlen(s) + 1;
      if (col > 77)
      {
	col = strlen(s)-1;
	putchar('\n');
      }
      else
	fputs("  ", stdout);
      fputs(s+1, stdout);
    }
  }
  if (!need_banner)
    putchar('\n');
  return 0;
}
