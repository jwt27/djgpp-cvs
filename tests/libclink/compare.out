Symbols in old library, but not in new one:
_GLOBAL_$I$_gppconio_init  __atexit  __mb_cur_max  _cleanup  _ctype_
_findenv  _findiop  _get_default_drive  _gppconio_init  _interrupt_stack_size
_iob  _memcpy  _setmode  _setstack  _smallbuf  _stklen  _tztab  fchmod
fcloseall  fgetpwent  longjmperror  nlist  on_exit  readcr  readv
realloc_srchlen  sigsetmask  stat_assist  timezone  tzsetwall  unsetenv
valloc  writecr  writev
